internal/health/health.go

```diff
@@ -0,0 +1,182 @@
+	h.readies = make(map[string]bool)
+	h.done = make(chan struct{})
+	go h.ticker()
+	return nil
+}
+
+func (h *Health) Stop() error {
+	close(h.done)
+	return nil
+}
+
+func (h *Health) ticker() {
+	fmt.Println("starting ticker")
+	tick := h.Clock.NewTicker(TickerTime)
+	for {
+		select {
+		case <-tick.Chan():
+			fmt.Println("got a tick")
+			h.mut.Lock()
+			for source, timeLeft := range h.timeLeft {
+				// only decrement positive counters since 0 means we're dead
+				if timeLeft > 0 {
+					fmt.Println("Decrementing timeout for source", source)
+					h.timeLeft[source] -= TickerTime
+					if h.timeLeft[source] < 0 {
+						h.timeLeft[source] = 0
+					}
+				}
+			}
+			h.mut.Unlock()
+		case <-h.done:
+			return
+		}
+	}
+}
+
+// Register a service with the health system. The timeout is the maximum
+// expected interval between service reports. If Ready is not called within that
+// interval (beginning from the time of calling Ready for the first time), it
+// (and the entire server) will be marked as not alive.
+func (h *Health) Register(source string, timeout time.Duration) {
+	h.mut.Lock()
+	defer h.mut.Unlock()
+	h.timeouts[source] = timeout
+	h.readies[source] = false
+	// we use a negative value to indicate that we haven't seen a report yet so
+	// we don't return "dead" immediately
+	h.timeLeft[source] = -1
+}
+
+// Ready is called by services to indicate their readiness to receive traffic.
+// If any service is not ready, the system as a whole is not ready.
+// Even unready services will be marked as alive as long as they report in.
+func (h *Health) Ready(source string, ready bool) {
+	h.mut.Lock()
+	defer h.mut.Unlock()
+	h.readies[source] = ready
+	h.timeLeft[source] = h.timeouts[source]
+}
+
+// Report returns the current health status of the system as a pair of booleans.
+// Alive is true at startup; once all services have reported in at least once, it
+// is true only if all services have reported within their timeout interval.
+func (h *Health) Report() (alive bool, ready bool) {
+	h.mut.RLock()
+	defer h.mut.RUnlock()
+	// if any counter is 0, we're dead
+	alive = true
+	for name, a := range h.timeLeft {
+		fmt.Printf("%s=%d\n", name, a)
+		alive = alive && a != 0
+	}
+	if !alive {
+		// can't be ready if we're not alive
+		return false, false
+	}
+	for name, r := range h.readies {
+		// can't be ready if any service has not reported yet
+		if h.timeLeft[name] < 0 {
+			return true, false
+		}
+		ready = ready || r
+	}
+	return alive, ready
+}
+
+func (h *Health) IsAlive() bool {
+	h.mut.RLock()
+	defer h.mut.RUnlock()
+	// if any counter is 0, we're dead
+	for _, a := range h.timeLeft {
+		if a == 0 {
+			return false
+		}
+	}
+	return true
+}
+
+func (h *Health) IsReady() bool {
+	h.mut.RLock()
+	defer h.mut.RUnlock()
+	// if any counter is not positive, we're not ready
+	for _, a := range h.timeLeft {
+		if a <= 0 {
+			return false
+		}
+	}
+
+	// if no one has registered yet, we're not ready
```

Comment:
If no one has registered, then `h.timeLeft` will be all `0` and that should be covered by line 166?

Reply:
If no one has registered, these loops don't get run, so line 166 never gets executed.